### ISSUE TITLE
Image is not re-tagged if tag was removed since last build

### ISSUE THREAD

**Issue by sschuberth** [bmuschko/gradle-docker-plugin#697]
When building an image via the `DockerBuildImage` task with a `tag` configured, and manually deleting that tag via `docker image rm <tag>`, re-running the tasks says it's up-to-date, not re-creating the tag.

### Expected Behavior

Given the task (in a Java application sub-project called "cli")

    task dockerBuildBaseImage(type: com.bmuschko.gradle.docker.tasks.image.DockerBuildImage) {
        description = 'Builds the base Docker image to run ORT.'
        // This uses 'docker/Dockerfile'.
        tag = 'ort-base:latest'
    }

and running

$ ./gradlew cli:dockerBuildBaseImage
$ docker image rm ort-base:latest
$ ./gradlew cli:dockerBuildBaseImage

I'd expect the tag to be re-created.

### Current Behavior

The tag is not re-created and the task is displayed as "up-to-date".

### Context

I accidentally deleted the tag and wanted to re-create it. As I didn't know the hash, re-running the task seemed like an easy way to re-create it.

### Steps to Reproduce (for bugs)

See expected behavior above.

### Your Environment

Ubuntu 18.04.

**Comment by bmuschko** [comment]
That's currently working as designed. However, I am aware of the described shortcoming. I also mentioned the issue [here](https://github.com/bmuschko/gradle-docker-plugin/issues/686).

I didn't implement a call to the Docker registry yet because additional refactorings are required. The Docker client needs to be available in the constructor instead of the action. At the moment it's only available in the action. The workaround is to run the specific clean task e.g. `cli:cleanDockerBuildBaseImage`. You don't really need to clean everything.

If more users run into the issue, then we'll have to think about reverting the change. Alternatively, I'd be happy to accept a PR that implements the desired behavior.

**Comment by sschuberth** [comment]
Thanks for the explanation. Do you have any feeling whether these refactorings would be easier in the context of doing #688?

**Comment by Vampire** [comment]
I run into the same issue basically.
I played with a Gradle build that builds a new image and then updates the service for continuous deployment of a chat-bot.
For this I basically changed one setting in the docker file back and forth and then wondered that it doesn't work.
It didn't work because the the task was taken from task-output-cache which is of course not correct as the output is not only the file with the id, but also the actual tagged image on the docker engine.

I guess just doing a proper up-to-date check that checks whether the tag exists in the docker engine and points to the right id will not be sufficient, iirc the task also has to be made non-cacheable.

**Comment by Vampire** [comment]
Well, the up to date check can make a task out of date even if it would be up to date otherwise. How should this work with the task output cache. In this concrete example the up to date check would probably check that the tag exists and points to the Id stored in the output file. But to check this, you need the output file already.

The only way I can imagine how to use the build cache is to not use the build image task as it is, but have a task that produces from the same inputs an exported / saved image file and then have a task that loads and tags that image. The first task can be made cacheable easily, as it has only local file outputs that can be taken from the cache and are relocatable. The second task then again must not be cacheable, as its output is in the engine, but its up to date check again could check whether the tag is present and points to the correct id.

**Comment by sschuberth** [comment]
> a proper up-to-date check that checks whether the tag exists in the docker engine and points to the right id will not be sufficient

Sorry for my ignorance, but why not?

**Comment by cdancy** [comment]
We've discussed impls to this over the years now and I don't think we could ever really build a "complete" solution here and really the best we could do is a "best guess" from gradles point of view.

**Comment by orzeh** [comment]
Agreed. IMO it's good the change was reverted. There are many corner cases which make it difficult to provide reasonable solution.

**Comment by bmuschko** [comment]
@orzeh So if we'd make the task non-cacheable and would query Docker engine for an image with that specific ID what are the other corner cases you are thinking of?

**Comment by sschuberth** [comment]
> have a task that produces from the same inputs an exported / saved image file and then have a task that loads and tags that image.

Breaking things up into two tasks this way were my thoughts exactly.

**Comment by orzeh** [comment]
> So if we'd make the task non-cacheable and would query Docker engine for an image with that specific ID what are the other corner cases you are thinking of?

One thing that comes to my mind is that suppose you have a task which builds image with two tags: `latest` and `X.Y`. You release also tag `X.Y.Z` but it's just a side note. You use a base image, let's say `alpine`, wherein some vulnerability was discovered. You may want to release new version of your images even if your app code didn't change and tags exist in Docker engine. Hope that makes sense.

**Comment by bmuschko** [comment]
@orzeh Wouldn't the use of a slightly different base image create a new, unique image ID?

**Comment by orzeh** [comment]
@bmuschko yes. But does it change anything?

**Comment by bmuschko** [comment]
@orzeh My understanding is that we _can_ implement reliable incremental build functionality for the `DockerBuildImage` task with something like this:

```
DockerBuildImage(dockerClient) {
    outputs.upToDateWhen {
        // check if image with ID already exists in local registry
    }
}
```

You an @cdancy raised concerns that it would an approximation or that there are corner cases to be considered. I am just trying to figure out if the approach above is flawed.

**Comment by cdancy** [comment]
So if the image exists is it "newer" than the image we are about to build? As the docker daemon can be remote, and thus N number of people using it for whatever reasons, can we be sure the image we are building locally is again "newer" than what is already available.

**Comment by orzeh** [comment]
@bmuschko I think my previous post demonstrates that: you _may_ want to rebuild and tag image even if project binaries didn't changed and image with ID that you have saved during previous build exists in Docker engine.

I also not sure what do you mean by _"exists in local registry"_. Would you check Docker daemon or Docker registry?

**Comment by cdancy** [comment]
@orzeh @bmuschko it's a messy problem to solve and I don't know if there is a "clean" way to do so. Doesn't feel like there is as we've tried this a few times now over the years.

**Comment by orzeh** [comment]
@cdancy agreed. But if we can give users some building blocks to implement _up-to-date_ strategies on their own it would be valuable. Worth spending some time on discussion.

**Comment by cdancy** [comment]
> up-to-date strategies

@orzeh having different "strategies" folks could use for varying scenarios _may_ be a way forward. I could see users supplying an enum/string that we would interrogate for a given strategy to use.

**Comment by cdancy** [comment]
There is a detailed thread somewhere (probably multiples at this point) of us going back and forth trying various things. Might be worth putting together a new sort of spreadsheet with each of these as they arise to track all the possible paths. But even as I think about it the very act of doing so is telling me this may not be a problem for us to solve as there is too much out of our control. IDK.
